README.md
## Capitaine Vaudou

  A Capitaine Vaudou role-playing game framework for MapTool virtual tabletop software.

### Status

  * pending 2021-02-24

### Milestone

  * beta

### File System

  * doc -> documents (txt)
  * rtw -> release to web (cmpgn)
  * src -> sources (html, css, js, mt)

### Language

  For now, only **french** language is supported. Feel free to *pull request* or *contact repository owner* if you're interested in english support ; many hints and suggested translations are already available in /doc.

### License

  The **Capitaine Vaudou** game and universe are the legal property of *Jean-Pierre Pécau*.
  The **Simulacres** game system is the legal property of *Pierre Rosenthal*.
  Capitaine Vaudou is published by *Black Book Editions* and *Monolith*. © 2020

  The **MapTool** framework, directed by *Kleey*, coded by *Zedafty* and tested by *Jambon*, is released on *GitHub* under *Creative Commons* license **BY** and **NC** (Attribution and Non commercial).


doc/Changes.txt
================================================================================
--------------------------------------------------------------------------------
# Capitaine Vaudou MapTool Framework : Changes
--------------------------------------------------------------------------------
================================================================================

	24/02/2021 : Bêta 1.0
	16/01/2021 : Alpha 3.3
	14/12/2020 : Alpha 3.2
	09/12/2020 : Alpha 3.1
	28/11/2020 : Alpha 3.0
	23/11/2020 : Alpha 2.3
	21/11/2020 : Alpha 2.2
	20/11/2020 : Alpha 2.1
	20/11/2020 : Alpha 2.0
	17/11/2020 : Alpha 1
	14/11/2020 : Draft

================================================================================
--------------------------------------------------------------------------------
# Latest : Bêta 1.1
--------------------------------------------------------------------------------
================================================================================

TODO : Ajouter un Compte-tours pour le MJ
TODO : Permettre la réorganisation de l'inventaire sur la dernière ligne
TODO : Voir pour créer une icône d'activation des EP + des icônes des états (héroïsme + 3 énergies de base)
TODO : Voir pour renverser la logique de la consommation des énergies
IDEA : Check change over character sheet by putting event onMouseEnter() followed by a simple that checks if the player has an assigned token
! max pour l'inventaire des personnages => pas de limite
! max pour l'inventaire des navires => pas de limite

================================================================================
--------------------------------------------------------------------------------
# Latest : Bêta 1.0
--------------------------------------------------------------------------------
================================================================================

# Campagne
° Mettre à jour les propriétés de campagne (navire uniquement)
° Ajouter la prise en charge des touches de contrôle pour les champs numériques (Shift et Ctrl)
° Harmoniser les schémas de couleurs (PJ = marron/rouge ; PNJ = vert-gris/turquoise)
° Réorganiser les icônes (changement position et amélioration des vecteurs de base)
° Ajouter un historique pour les personnages et les bateaux (sous l'en-tête)
° Ajouter un widget pour la météo (température, vent et ciel)

# Fiche de personnage
° Définir un minimum de 3 pour les Composantes et de 1 pour les Moyens

# Fiche de navire
° Ajouter des menus déroulants pour les valeurs des maîtres d'équipage
° Ajouter des boutons avec icône pour les boulets de canons (creux, ramés, mitraille et rouges)
° Ajouter des boutons avec icône pour les jets de dés (tests de navigation, manoeuvre et canonnade)
° Ajouter la prise en charge de la distance pour les jets de canonnade
° Ajouter l'affichage de la cadence de tir pour les jets de canonnade
° Ajouter des indicateurs de conditions de canonnade impossible
° Changer le nombre maximal de boulets à bord (60 boulets par canon au total)
° Ajouter le bonus à la marge de réussite pour les canonnades
° Ajouter les dégâts principaux et de jumelage pour les cannonades
° Ajouter la cible pour les tests de canonnade
° Ajouter un modificateur de circonstances pour les tests de manoeuvre

# Lanceur de dés
° Limiter le nombre de cible à seulement une (liste de choix plutôt que cases à cocher)
° Ajouter des astuces au survol des éléments (dans le titre de section) plutôt que des infobulles
° Ajouter le choix d'une combinaison en cliquant individuellement sur les composants et les moyens
° Ajouter +1 au Seuil pour chaque Potentiel de base activé (Puissance, Eapidité et Précision)
° Afficher le nombre de points dépensés dans chaque Potentiel lors d'un jet de dés (sous l'icône)

# Bugs
° Corriger la formulation erronée de cadence de tir (3 tirs par tour => tous les 3 tours)
° Corriger la sortie du résultat de jet de dés dans la fenêtre de discussion (console.log)
° Corriger l'affichage des informations de licence (chaque nouveau joueur connecté uniquement)
° Corriger l'erreur d'interprétation du résultat de jet de dés (réussite si jet égal au seuil)

================================================================================
--------------------------------------------------------------------------------
# Changes : Alpha 3.3
--------------------------------------------------------------------------------
================================================================================

# Campagne
° Modifier le jeu de propriétés 'Basic' (par défaut les pions sont des objets sans propriété)
° Modifier les propriétés de personnage et de navire (changement de casse et de noms)
° Réorganiser les librairies (renommage, ordonnancement, création et suppression)
° Améliorer la prise en charge des caractères spéciaux des champs de saisie
° Ajouter des constantes globales (librairie Lib:Global macro getConst)
° Ajouter des erreurs globales (librairie Lib:Global macro getError)
° Ajouter des macros de gestion des pions (fenêtre MJ dans MapTool)
° Ajouter la prise en charge des pions de joueur (i.e. propriétés globalement accessibles)

# Barre d'outils
° Renommer la surcouche 'Overlay' en 'Toolbar'
° Ajouter la possibilité de placer la barre d'outils en bas au centre

# Fiche de personnage
° Remplacer fonctions de chaque propriété de pions par une fonction générique
° Remplacer fonction getName() par getCharName() (i.e. nom de personnage plutôt que nom de pion)
° Renommer état 'Fou' (insane) en 'Crise de nerfs'  (nervous_breakdown)
° Ajouter Talent X Astronomie (tests de Navigation nocturne)

# Fiche de navire
° Ajouter la fiche de navire
° Changer la méthode de gestion de l'équipage (seul le nombre total de membres actuels est saisi)
° Inverser les valeurs de navir de la Frégate et du Galion (erreur suspectée dans les règles)
° Permuter la valeur d'équipage maximale du navire Marchand de 45 à 80 (erreur suspectée dans les règles)

src/2. Frames/3. JS/Char.js
[r: "
/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/*                                                                            */
/*                   CAPITAINE VAUDOU CHARACTER JAVASCRIPT                    */
/*                                                                            */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Pools (front-end only) */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function checkPools() {
	let hp = document.getElementById('hp');
	let bp = document.getElementById('bp');
	let mp = document.getElementById('mp');
	let o = document.querySelector('.portrait').parentElement;
	bp.value == 0 ? o.classList.add('breathless') : o.classList.remove('breathless');
	mp.value == 0 ? o.classList.add('nervous-breakdown'): o.classList.remove('nervous-breakdown');
	if (hp.value == 0) {
		o.classList.add('dead');
		o.classList.remove('breathless');
		o.classList.remove('nervous-breakdown');
		o.classList.remove('unconscious');
	} else {
		o.classList.remove('dead');
	}
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Global Value Functions */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function updateGlobalValue(e) { // e = HTML element
	let min = parseInt(e.getAttribute('min'));
	let max = parseInt(e.getAttribute('max'));
	let v = Math.min(Math.max(e.value, min), max); // 2, 4, 6 , 8, 10
	let m = 6, n = 4; // max values
	if (v < 10) m = n = Math.floor(v / 2); // 1, 2, 3, 4, 5
	e.value = v;
	callTimedFunction(function() {
		callMacro('jobj=Components;body=' + m);
		callMacro('jobj=Components;heart=' + m);
		callMacro('jobj=Components;mind=' + m);
		callMacro('jobj=Means;perception=' + n);
		callMacro('jobj=Means;action=' + n);
		callMacro('jobj=Means;resistance=' + n);
		refreshOverlay('DiceRoller'); // anyway, refresh dice roller overlay (time differed)
	});
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Abilities (front-end only) */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

////////////////////////////////////////////////////////////////////////////////
// * Add Ability (front-end only)
////////////////////////////////////////////////////////////////////////////////

function addAbility(p) { // p = HTML element
	let o = document.querySelector('#abilities .block');
	let max = 5;
	let num = o.children.length - 1;
	if (num < max) {
		let e = document.createElement('INPUT');
		e.setAttribute('type', 'text');
		e.setAttribute('placeholder', 'Aptitude');
		e.setAttribute('maxlength', 24);
		attachAbilityEvent(e);
		o.insertBefore(e, p);
		callMacro('', 'addAbility'); // append new item in abilities array
		if (num + 1 == max) {
			p.classList.add('hidden');
			p.dataset.hidden = 'true';
		}
	}
}

////////////////////////////////////////////////////////////////////////////////
// * Edit Ability (front-end only)
////////////////////////////////////////////////////////////////////////////////

function updateAbility(e) { // e = HTML element
	if (document.getElementById('abilities').dataset.locked == 'true') return;
	let p = document.getElementById('add_ability');
	let v = encodeURIComponent(replaceIllegalChars(e.value.trim()));
	let i = -1, j = 0;
	document.querySelectorAll('#abilities input').forEach(function(o) {
		if (e == o) i = j;
		j++;
	});
	if (v.length == 0) { // delete ability
		e.remove();
		callMacro(i, 'deleteAbility'); // delete item at index
		if (p.dataset.hidden == 'true') {
			p.classList.remove('hidden');
			delete p.dataset.hidden;
		}
	} else { // edit ability
		callMacro('index=' + i + ';value=' + v, 'editAbility'); // edit item at index
	} refreshOverlay('DiceRoller'); // anyway, refresh dice roller overlay (time differed)
}

////////////////////////////////////////////////////////////////////////////////
// * Attach Ability Event
////////////////////////////////////////////////////////////////////////////////

function attachAbilityEvent(e) { // e = HTML element
	e.addEventListener('focus', function() {
		this.dataset.focus = 'true';
	});
	e.addEventListener('change', function() {
		delete this.dataset.focus;
		updateAbility(this);
	});
	e.addEventListener('blur', function() {
		if (this.dataset.focus == 'true') updateAbility(this);
	})
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Skills Functions */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function hideEmptySkills() {
	document.querySelectorAll('#skills table input').forEach(function(o) {
		if (o.id != 'cold_weapon' && o.id != 'cold_weapon_favored' && (o.value == '' || o.value == 'nil')) o.parentElement.parentElement.style.display = 'none';
	});
}

function showEmptySkills() {
	document.querySelectorAll('#skills table tr').forEach(function(o) {
		o.style.display = '';
	});
}

function toggleEmptySkills(b) { // b = hidden flag
	b ? hideEmptySkills() : showEmptySkills();
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Document Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

// * Purse
document.getElementById('purse_submit').addEventListener('click', function() {
	document.getElementById('purse').click();
});
document.getElementById('purse').addEventListener('click', function() {
	if (this.checked) document.getElementById('purse_amount').value = 0;
});

// * Pools
document.querySelectorAll('#pools input').forEach(function(o) {
	o.addEventListener('change', function() {
		let max = this.id.length == 6 ? this : null;
		let cur = this.id.length == 2 ? this : null;
		if (max) { // is maximum value
			updateTokenProperty(max, 'Pools');
			let n = max.value;
			let k = max.id.substr(0, 2);
			let q = document.getElementById(k);
			if (parseInt(q.value) > parseInt(n)) {
				q.value = n;
				cur = q;
			} q.max = n;
		}
		if (cur) { // is current value OR current value has changed
			updateTokenProperty(cur, 'Pools');
			callTimedFunction(function() {
				callMacro('false', 'checkPools'); // check and update token pools without refreshing sheet (back-end)
				checkPools(); // check and update displayed pools in sheet (front-end)
			});
		}
	});
});

// * Components
document.querySelectorAll('#components input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, 'Components');
	})
});

// * Means
document.querySelectorAll('#means input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, 'Means');
	})
});

// * Global Value
document.querySelector('#global_value input').addEventListener('change', function() {
	updateGlobalValue(this);
});

// * Targets
document.querySelectorAll('#targets input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, 'Targets');
	})
});

// * Energies
document.querySelectorAll('#energies input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, 'Energies');
	})
});

// * Potentials
document.querySelectorAll('#potentials input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, 'Potentials');
	})
});

// * Skills
document.querySelectorAll('#skills table input').forEach(function(o) {
	o.addEventListener('change', function() {
		updateTokenProperty(this, this.id == 'cold_weapon_favored' ? 'Char' : 'Skills');
	});
});

// * Abilities
document.getElementById('add_ability').addEventListener('click', function() {
	addAbility(this);
});
document.querySelectorAll('#abilities input').forEach(function(o) {
	attachAbilityEvent(o);
});

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Window Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

window.addEventListener('focus', function() {
	callMacro('', 'checkSheet', 'Scripts');
});
"]


src/2. Frames/3. JS/Tool.js
[r: "
/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/*                                                                            */
/*                    CAPITAINE VAUDOU TOOLBAR JAVASCRIPT                     */
/*                                                                            */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Configuration */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

var cfg = {
	'bubble' : {
		'delay' : 1500, // Time in milliseconds during which the bubbling icons will remain visible until hovered ; Default : 1500
		'pinch' : 150, // Time in milliseconds before the bubbling is hidden on mouse click ; Default : 150
		'x_adj' : -30, // Bubbling icon horizontal adjustement in pixels regarding of cursor ; Default : -32
		'y_adj' : -30 // Bubbling icon position adjustement in pixels regarding of cursor ; Default : -32
	},
	'weather' : {
		'reload' : true, // Force all connected players toolbar to be reloaded automatically ; Default : true
		'degree' : {
			'raz' : 20, // Reinitialization temperature degree (celsius) ; Default : 20
			'max' : 45, // Maximal temperature degree (celsius) ; Default : 45
			'min' : -10, // Minimal temperature degree (celsius) ; Default : -10
			'delay' : 150, // Time elasped between two update degree occurences ; Default : 150
			'speed' : {
				'slow' : 4, // Number of update degree occurences before speeding up to soft ; Default : 4
				'soft' : 8, // Number of update degree occurences before speeding up to soft ; Default : 8
				'fast' : 16 // Number of update degree occurences before speeding up to soft ; Default : 16
			}
		}
	}
};

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Languages */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

var lng = {
	// * Wind
	'east'         : 'Est',
	'south_east'   : 'Sud-Est',
	'south'        : 'Sud',
	'south_west'   : 'Sud-Ouest',
	'west'         : 'Ouest',
	'north_west'   : 'Nord-Ouest',
	'north'        : 'Nord',
	'north_east'   : 'Nord-Est',
	// * Sky
	'empty'        : '',
	'sunny'        : 'Ensoleill&eacute;',
	'veiled'       : 'Voil&eacute;',
	'cloudy'       : 'Nuageux',
	'fog'          : 'Brouillard',
	'drizzle'      : 'Bruine',
	'covered'      : 'Couvert',
	'rainy'        : 'Pluvieux',
	'stormy'       : 'Orageux'
};

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Graphical User Interface */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

var gui = {
	'timeout' : null,
	'bubbled' : false,
	'mouse' : {
		'up' : false
	},
	'weather' : {
		'degree' : {
			'delay' : 150,
			'down' : false,
			'interval' : null,
			'count' : 0
		},
		'wind' : {
			'down' : false
		}
	}
};

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Call Functions */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function callMacro(val, src, lib, out, target) {
	if (src === undefined) src = 'setProp';
	if (lib === undefined) lib = 'Scripts';
	let url = 'macro://' + src + '@Lib:' + lib + '/' + (out === undefined ? 'none' : out) + '/' + (target === undefined ? '' : target) + '?' + val;
	let o = document.getElementById('macrolink');
	// ---------------------------------------------------------------------
	// * MapTool
	// ---------------------------------------------------------------------
	o.href = url;
	o.click();
	// ---------------------------------------------------------------------
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Bubbling Functions */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function startBubblingTimeout(n) {
	if (typeof n !== 'number') n = cfg.bubble.delay;
	let o = document.getElementById('bubble');
	if (!gui.bubbled) {
		gui.bubbled = true;
		setTimeout(function() {
			o.style.display = '';
		}, cfg.bubble.pinch);
	}
	gui.timeout = setTimeout(function() {
		o.style.display = 'none';
	}, n);
}

function stopBubblingTimeout() {
	clearTimeout(gui.timeout);
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Bubbling Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function attachBubblingEvents() {
	document.getElementById('bubble').addEventListener('mouseenter', stopBubblingTimeout);
	document.getElementById('bubble').addEventListener('mouseleave', startBubblingTimeout);
	document.addEventListener('mouseup', function(e) {
		if (!gui.mouse.up) {
			let o = document.getElementById('bubble');
			o.style.left = (e.clientX - o.offsetWidth + cfg.bubble.x_adj) + 'px';
			o.style.top = (e.clientY - o.offsetHeight + cfg.bubble.y_adj) + 'px';
			o.style.opacity = '1';
			gui.mouse.up = true;
			startBubblingTimeout();
		}
	});
	document.addEventListener('mousedown', function() {
		startBubblingTimeout(cfg.bubble.pinch);
	});
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Weather Functions */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function udpateWeatherCaption(s, p) { // s = string, p = HTML element (optional)
	if (s === undefined) s = '';
	let o = document.querySelector('.weather > .caption')
	if (s == '') o.classList.add('hidden');
	else if (p !== undefined) {
		o.classList.remove('hidden');
		o.querySelector('span').innerHTML = s;
		o.style.top = p.offsetTop - o.offsetHeight;
		o.style.left = p.offsetLeft + Math.round(p.offsetWidth / 2) - Math.round(o.offsetWidth / 2);
	}
}

function getDegree() { // returns integer
	let o = document.querySelector('.weather .degree output');
	return parseInt(o.innerText);
}

function setDegree(i) { // i = integer
	let o = document.querySelector('.weather .degree output');
	o.innerText = i;
}

function updateDegree(d) { // d = direction (signed integer)
	let p = document.querySelector('.weather .degree');
	let o = p.querySelector('output');
	let n = cfg.weather.degree.delay;
	let b = true;
	o.innerText = Math.max(Math.min(parseInt(o.innerText) + 1 * d, cfg.weather.degree.max), cfg.weather.degree.min);
	if (gui.weather.degree.count == cfg.weather.degree.speed.slow) {
		p.classList.add('slow');
		n /= 2;
	} else if (gui.weather.degree.count == cfg.weather.degree.speed.soft) {
		p.classList.add('soft');
		n /= 4;
	} else if (gui.weather.degree.count == cfg.weather.degree.speed.fast) {
		p.classList.add('fast');
		n /= 10;
	} else {
		b = false;
	}
	gui.weather.degree.count++;
	if (b) {
		clearInterval(gui.weather.degree.interval);
		gui.weather.degree.interval = setInterval(function() {
			updateDegree(d);
		}, Math.round(n));
	}
}

function rotateWindArrow(r) { // r = radian angle
	let o = document.querySelector('.weather .wind .arrow');
	o.style.transform = 'rotate(' + (r + Math.PI / 2) + 'rad)';
}

function updateWindDirection(r) { // r = radian angle
	let o = document.querySelector('.weather .wind');
	let s = '';
	let n = Math.PI;
	if (r < 0) {
		if (r <= -n * 7/8) s = lng.west;
		else if (r <= -n * 5/8) s = lng.north_west;
		else if (r <= -n * 3/8) s = lng.north;
		else if (r <= -n * 1/8) s = lng.north_east;
		else s = lng.east;
	} else {
		if (r <= n * 1/8) s = lng.east;
		else if (r <= n * 3/8) s = lng.south_east;
		else if (r <= n * 5/8) s = lng.south;
		else if (r <= n * 7/8) s = lng.south_west;
		else if (r <= n) s = lng.west;
	} o.dataset.title = s;
}

function updateWindArrow(e) { // e = mouse event ; returns radian angle
	let p = document.querySelector('.weather .wind');
	let b = p.getBoundingClientRect();
	let x = e.clientX - Math.floor(b.x + (b.width / 2));
	let y = e.clientY - Math.floor(b.y + (b.height / 2));
	let r = Math.atan2(y, x);
	rotateWindArrow(r);
	updateWindDirection(r);
	udpateWeatherCaption(p.dataset.title, p);
	return r;
}

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Weather Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

function attachWeatherEvents() {
	document.querySelector('.weather .degree').addEventListener('contextmenu', function(e) {
		e.preventDefault(); // no context menu on right click
	});
	document.querySelector('.weather .degree').addEventListener('mousedown', function(e) {
		if (!gui.weather.degree.down) {
			gui.weather.degree.down = true;
			this.classList.add('active');
			let i = e.button == 0 ? 1 : -1; // positive if left click, negative otherwise
			if (e.shiftKey) setDegree(cfg.weather.degree[(i > 0 ? 'max' : 'min')]);
			else if (e.ctrlKey) setDegree(cfg.weather.degree.raz);
			else {
				updateDegree(i);
				gui.weather.degree.interval = setInterval(function() {
					updateDegree(i);
				}, gui.weather.degree.delay);
			}
		}
	});
	document.querySelector('.weather .wind').addEventListener('mousedown', function() {
		if (!gui.weather.wind.down) {
			gui.weather.wind.down = true;
			this.classList.add('active');
		}
	});
	document.addEventListener('mouseup', function(e) {
		if (gui.weather.degree.down) {
			gui.weather.degree.down = false;
			gui.weather.degree.count = 0;
			clearInterval(gui.weather.degree.interval);
			document.querySelector('.weather .degree').classList.remove('active', 'slow', 'soft', 'fast');
			callMacro('key=degree;val=' + getDegree(), 'setLibProp');
			if (cfg.weather.reload) callMacro('', 'reloadTool');
		}
		if (gui.weather.wind.down) {
			let r = updateWindArrow(e);
			gui.weather.wind.down = false;
			document.querySelector('.weather .wind').classList.remove('active');
			if (e.target.nodeType == 1 && !e.target.classList.contains('wind') && !e.target.classList.contains('arrow')) udpateWeatherCaption();
			callMacro('key=wind;val=' + r, 'setLibProp');
			if (cfg.weather.reload) callMacro('', 'reloadTool');
		}
	});
	document.addEventListener('mousemove', function(e) {
		if (gui.weather.degree.down) gui.weather.degree.x1 = e.clientX;
		if (gui.weather.wind.down) updateWindArrow(e);
	});
	document.querySelector('.weather .sky').addEventListener('click', function() {
		let o = document.querySelector('.weather .skyset');
		o.classList.contains('hidden') ? o.classList.remove('hidden') : o.classList.add('hidden');
	});
	document.addEventListener('mousedown', function(e) {
		let o = document.querySelector('.weather .skyset');
		if (!e.target.classList.contains('sky')) o.classList.add('hidden');
	});
	document.querySelectorAll('.weather .skyset li').forEach(function(o) {
		let q = document.querySelector('.weather .sky');
		o.addEventListener('mousedown', function() {
			q.innerHTML = o.innerHTML;
			if (o.dataset.sky !== undefined) q.dataset.title = lng[o.dataset.sky];
			callMacro('key=sky;val=' + o.dataset.sky, 'setLibProp');
			if (cfg.weather.reload) callMacro('', 'reloadTool');
		});
		o.addEventListener('mouseenter', function() {
			if (o.dataset.sky !== undefined) udpateWeatherCaption(lng[o.dataset.sky], o);
		});
	});
}

document.querySelectorAll('.weather .wind, .weather .sky').forEach(function(o) {
	o.addEventListener('mouseenter', function() {
		udpateWeatherCaption(o.dataset.title, o);
	});
});

document.querySelectorAll('.weather .wind, .weather .sky, .weather .skyset').forEach(function(o) {
	o.addEventListener('mouseleave', function() {
		udpateWeatherCaption();
	});
});

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Change Map Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

document.addEventListener('mousedown', function(e) {
	if (e.button == 0) { // left click
		callMacro('" + getCurrentMapName() + "', 'checkMap');
	}
});

/* ========================================================================== */
/* -------------------------------------------------------------------------- */
/* # Window Events */
/* -------------------------------------------------------------------------- */
/* ========================================================================== */

window.addEventListener('load', function() {
	let o = document.querySelector('.weather .wind');
	let p = document.querySelector('.weather .sky')
	// ---------------------------------------------------------------------------
	// * Dev
	// ---------------------------------------------------------------------------
	// let q = document.querySelector(`[data-sky='empty']`);
	// let r = -Math.PI * 4/8;
	// ---------------------------------------------------------------------------
	// * MapTool
	// ---------------------------------------------------------------------------
	let q = document.querySelector(`[data-sky='" + getLibProperty("sky", "Lib:Vars") + "']`);
	let r = parseFloat(" + getLibProperty("wind", "Lib:Vars") + ");
	if (r == '') r = -Math.PI * 4/8;
	// ---------------------------------------------------------------------------
	if (o.dataset.title === undefined) {
		rotateWindArrow(r);
		updateWindDirection(r);
	}
	if (p.dataset.title === undefined) {
		p.innerHTML = q.innerHTML;
		p.dataset.title = lng[q.dataset.sky];
	}
	if (document.getElementsByTagName('body')[0].classList.contains('gm')) { // gm only
		attachBubblingEvents();
		attachWeatherEvents();
	}
});
"]
